connmgr/banmanager.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"net"
+	"sync"
+
+	"github.com/decred/dcrd/peer/v3"
+)
+
+// BanManagerConfig is the configuration struct for the ban manager.
+type BanManagerConfig struct {
+	// DisableBanning represents the status of disabling banning of
+	// misbehaving peers.
+	DisableBanning bool
+
+	// BanThreshold represents the maximum allowed ban score before
+	// misbehaving peers are disconnecting and banned.
+	BanThreshold uint32
+
+	// BanPeer represents a banning callback for cleanup and status
+	// updates purposes on a peer ban.
+	BanPeer func(p *peer.Peer)
+
+	// MaxPeers indicates the maximum number of inbound and outbound
+	// peers allowed.
+	MaxPeers int
+
+	// Whitelist represents the whitelisted IPs of the server.
+	WhiteList []net.IPNet
+}
+
+// banMgrPeer extends a peer to maintain additional state maintained by the
+// ban manager.
+type banMgrPeer struct {
+	*peer.Peer
+
+	isWhitelisted bool
+}
+
+// BanManager represents a peer ban score tracking manager.
+type BanManager struct {
+	cfg       BanManagerConfig
+	peers     map[string]*banMgrPeer
```

Comment:
This should be keyed by `*peer.Peer` like `netsync.SyncManager.peers`.

Then, as also done in `netsync`, you could have a nice helper method for looking up the peer along with a warning if needed like so:

```Go
// lookupPeer returns the ban manager peer that maintains additional state for
// a given base peer.  In the event the mapping does not exist, a warning is
// logged and nil is returned.
//
// This function MUST be called with the ban manager mutex locked (for reads).
func (bm *BanManager) lookupPeer(p *peer.Peer) *banMgrPeer {
	bmp, ok := peers[p]
	if !ok {
		log.Warnf("Attempt to lookup unknown peer %s\nStack: %v", p,
			string(debug.Stack()))
		return nil
	}
	return bmp
}
```